core/src/main/java/org/apache/iceberg/io/DeleteSchemaUtil.java

```diff
@@ -20,13 +20,38 @@
 package org.apache.iceberg.io;
 
 import org.apache.iceberg.MetadataColumns;
+import org.apache.iceberg.Partitioning;
 import org.apache.iceberg.Schema;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.types.TypeUtil;
 import org.apache.iceberg.types.Types;
 
 public class DeleteSchemaUtil {
   private DeleteSchemaUtil() {
   }
 
+  public static Schema metadataTableSchema(Table table) {
+    return metadataTableSchema(table, Partitioning.partitionType(table));
+  }
+
+  public static Schema metadataTableSchema(Table table, Types.StructType partitionType) {
+    Schema result = new Schema(
+        MetadataColumns.DELETE_FILE_PATH,
```

Comment:
Are these real metadata columns or just columns that happen to have the same name? I don't think we should reuse column IDs as metadata columns are using reserved ones.

Reply:
One of the tricky things is to pick a column field id, because here it may conflict with the user field id.  (the row of position deletes is a exact struct of the user's schema).  So I will still put these constants on the class MetadataColumns, but as reserved columns instead of metadata columns.

Reply:
Actually looking into this, if I don't re-use the metadata column I have to change a lot of codes, like:

- BaseDataReader.constantsMap will need new method to populate partition based on new partition id, which I'll have to pass in based on what field is selected.  (Currently since both id's are the same it works)
- All the error checks for whether selected columns are in metadata columns need to take this into account.  Currently this uses MetadataColumns.metadataFieldIds(), which will need to be now something like MetadataColumns.metadataAndConstantFieldIds()  to take into account new column.

Reply:
Did the above, feel free to see if it is better or if its messier.

Reply:
Also adding some background:  I had tried earlier to use the existing metadata column for this table (which is populated), but then pushdown does not work for metadata columns.    It seems the pruneSchema() function is called after filter() function on SparkScanBuilder, so that the binding in filter will fail.

Reply:
Update, this should be resolved, do not re-use metadata column ids anymore.